spec.bs

```diff
@@ -44,19 +58,89 @@ This specification extends the client-side storage available beyond cookies.
 
-<h3 id="the-document-object">Changes to {{Document}}</h3>
+<div class=example>
 
-TBD
+Alex visits `https://social.example/`. The page sets a some local storage. This local storage has been set in a [=first-party-site context=].
 
-<h3 id="storage">Changes to various client-side storage mechanisms</h3>
+```javascript
+window.localStorage.setItem("userid", "1234");
+```
 
-TBD
+Later on, Alex visits `https://video.example/`, which has an <{iframe}> on it which loads `https://social.example/heart-button`. In this case, the `social.example` {{Document}} |doc| is in a [=third party context=], and the local storage set previously might or might not be visible depending on User Agent storage access policies.
 
-<h4 id="cookies">Cookies</h4>
+Script in the <{iframe}> can call |doc|`.`{{Document/requestStorageAccess(types)}} to request access.
 
-TBD
+```javascript
+let handle = await document.requestStorageAccess({localStorage: true});
+let userid = handle.localStorage.getItem("userid");
+```
+
+</div>
+
+<h3 id="document-changes">Changes to {{Document}}</h3>
+
+<pre class="idl">
+dictionary StorageAccessTypes {
+  boolean all = false;
+  boolean cookies = false;
+};
+
+interface StorageAccessHandle {
+};
+
+partial interface Document {
+  Promise&lt;boolean> hasUnpartitionedCookieAccess();
+  Promise&lt;StorageAccessHandle> requestStorageAccess(StorageAccessTypes types);
+};
+</pre>
+
+When invoked on {{Document}} |doc|, the <dfn export method for=Document><code>hasUnpartitionedCookieAccess()</code></dfn> method must run these steps:
+
+1. Return the invocation of {{Document/hasStorageAccess()}}.
+
+Note:
+Now that {{Document/requestStorageAccess(types)}} <span class=allow-2119>can</span> be used to request [=unpartitioned data=] with or without specifically requesting cookies, it <span class=allow-2119>must</span> be made clear that {{Document/hasStorageAccess()}} only returns true if [=first-party-site context=] cookies are accessable to the current document.
+As a function name, {{Document/hasUnpartitionedCookieAccess()}} more clearly communicates this.
+For now {{Document/hasStorageAccess()}} is not considered deprecated, but that <span class=allow-2119>may</span> be worth taking up in future.
+
+When invoked on {{Document}} |doc|, the <dfn export method for=Document><code>requestStorageAccess(types)</code></dfn> method must run these steps:
+
+1. Let |p| be [=a new promise=].
+1. If |types|.{{StorageAccessTypes/all}} is `false` and |types|.{{StorageAccessTypes/cookies}} is `false`:
+    1. [=/Reject=] |p| with an "{{InvalidStateError}}" {{DOMException}}.
+    1. Return |p|.
+1. Let |requestUnpartitionedCookieAccess| be `true` if |types|.{{StorageAccessTypes/all}} is `true` or |types|.{{StorageAccessTypes/cookies}} is `true`, and `false` otherwise.
+1. Run the following steps [=in parallel=]:
```

Comment:
I don't think you can / should run this in parallel (which is the equivalent of moving to the browser process), the rSA algorithm will do its own in parallel once it gets there and you can't resolve / reject promises in parallel (would need to queue a task for that).

So I think you can just remove this, unless I'm missing something.